include/caffe/solver.hpp

```diff
@@ -159,6 +158,22 @@ class RMSPropSolver : public SGDSolver<Dtype> {
 };
 
 template <typename Dtype>
+class AdaDeltaSolver : public SGDSolver<Dtype> {
+ public:
+  explicit AdaDeltaSolver(const SolverParameter& param)
+      : SGDSolver<Dtype>(param) { PreSolve(); }
+  explicit AdaDeltaSolver(const string& param_file)
+      : SGDSolver<Dtype>(param_file) { PreSolve(); }
+
+ protected:
+  void PreSolve();
+  virtual void Regularize(int param_id);
```

Comment:
I didn't find a significant difference between your `AdaDeltaSolver::Regularize` and the original `SGDSolver::Regularize` (the former is almost duplicated from the latter).

We can just use `SGDSolver::Regularize` and avoid introducing a new regularization function in AdaDeltaSolver.